tests/test_histogram.py

```diff
@@ -110,3 +117,85 @@ def test_compare_histograms_optimized_vs_general():
+                                             binned_feature,
+                                             ordered_gradients,
+                                             ordered_hessians)
+
+    hist_naive = _build_ghc_histogram_naive(n_bins, sample_indices,
+                                            binned_feature, ordered_gradients,
+                                            ordered_hessians)
+
+    for hist in (hist_gc_root, hist_ghc_root, hist_gc, hist_gc, hist_ghc):
+        assert_array_equal(hist['count'], hist_naive['count'])
+        assert_allclose(hist['sum_gradients'], hist_naive['sum_gradients'])
+    for hist in (hist_ghc_root, hist_ghc):
+        assert_allclose(hist['sum_hessians'], hist_naive['sum_hessians'])
+    for hist in (hist_gc_root, hist_gc):
+        assert_array_equal(hist['sum_hessians'], np.zeros(n_bins))
+
+
+def test_hist_subtraction():
+    # Make sure the histogram subtraction trick gives the same result as the
+    # classical method.
+    rng = np.random.RandomState(42)
+    n_samples = 10
+    n_bins = 5
+    sample_indices = np.arange(n_samples).astype(np.uint32)
+    binned_feature = rng.randint(0, n_bins - 1, size=n_samples, dtype=np.uint8)
+    ordered_gradients = rng.randn(n_samples).astype(np.float32)
+    ordered_hessians = np.ones(n_samples, dtype=np.float32)
+
+    hist_parent = _build_ghc_histogram_unrolled(n_bins, sample_indices,
+                                                binned_feature,
+                                                ordered_gradients,
+                                                ordered_hessians)
+
+
+    mask = rng.randint(0, 2, n_samples).astype(np.bool)
+
+    sample_indices_left = sample_indices[mask]
+    ordered_gradients_left = ordered_gradients[mask]
+    ordered_hessians_left = ordered_hessians[mask]
+    hist_left = _build_ghc_histogram_unrolled(n_bins, sample_indices_left,
+                                              binned_feature,
+                                              ordered_gradients_left,
+                                              ordered_hessians_left)
+
+    sample_indices_right = sample_indices[~mask]
+    ordered_gradients_right = ordered_gradients[~mask]
+    ordered_hessians_right = ordered_hessians[~mask]
+    hist_right = _build_ghc_histogram_unrolled(n_bins, sample_indices_right,
+                                               binned_feature,
+                                               ordered_gradients_right,
+                                               ordered_hessians_right)
+
+    hist_left_sub = _subtract_ghc_histograms_unrolled(n_bins, hist_parent,
+                                                      hist_right)
+    hist_right_sub = _subtract_ghc_histograms_unrolled(n_bins, hist_parent,
+                                                      hist_left)
+
+    for key in ('count', 'sum_hessians', 'sum_gradients'):
+        assert_allclose(hist_left[key], hist_left_sub[key], rtol=1e-6)
```

Comment:
Can't you compare the full record array at once?

```python
assert_allclose(hist_left, hist_left_sub, rtol=1e-6)
```

or did you intentionally use fine grained comparison on each key individually to get more meaningful error reports in case of assertion failures?

Reply:
I tried but I'm getting this error:
<details>
```
    
>       assert_allclose(hist_left, hist_left_sub, rtol=1e-6)

tests/test_histogram.py:140: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../.virtualenvs/pygbm/lib/python3.7/site-packages/numpy/testing/_private/utils.py:1447: in compare
    equal_nan=equal_nan)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

a = array([(-0.61542976, 1., 1), (-2.1266346 , 2., 2), (-0.34148723, 2., 2),
       ( 2.2112024 , 3., 3), ( 0.        , 0., 0)],
      dtype=[('sum_gradients', '<f4'), ('sum_hessians', '<f4'), ('count', '<u4')])
b = array([(-0.61542976, 1., 1), (-2.1266346 , 2., 2), (-0.34148723, 2., 2),
       ( 2.2112024 , 3., 3), ( 0.        , 0., 0)],
      dtype=[('sum_gradients', '<f4'), ('sum_hessians', '<f4'), ('count', '<u4')])
rtol = 1e-06, atol = 0, equal_nan = True

    def isclose(a, b, rtol=1.e-5, atol=1.e-8, equal_nan=False):
        """
        Returns a boolean array where two arrays are element-wise equal within a
        tolerance.
    
        The tolerance values are positive, typically very small numbers.  The
        relative difference (`rtol` * abs(`b`)) and the absolute difference
        `atol` are added together to compare against the absolute difference
        between `a` and `b`.
    
        .. warning:: The default `atol` is not appropriate for comparing numbers
                     that are much smaller than one (see Notes).
    
        Parameters
        ----------
        a, b : array_like
            Input arrays to compare.
        rtol : float
            The relative tolerance parameter (see Notes).
        atol : float
            The absolute tolerance parameter (see Notes).
        equal_nan : bool
            Whether to compare NaN's as equal.  If True, NaN's in `a` will be
            considered equal to NaN's in `b` in the output array.
    
        Returns
        -------
        y : array_like
            Returns a boolean array of where `a` and `b` are equal within the
            given tolerance. If both `a` and `b` are scalars, returns a single
            boolean value.
    
        See Also
        --------
        allclose
    
        Notes
        -----
        .. versionadded:: 1.7.0
    
        For finite values, isclose uses the following equation to test whether
        two floating point values are equivalent.
    
         absolute(`a` - `b`) <= (`atol` + `rtol` * absolute(`b`))
    
        Unlike the built-in `math.isclose`, the above equation is not symmetric
        in `a` and `b` -- it assumes `b` is the reference value -- so that
        `isclose(a, b)` might be different from `isclose(b, a)`. Furthermore,
        the default value of atol is not zero, and is used to determine what
        small values should be considered close to zero. The default value is
        appropriate for expected values of order unity: if the expected values
        are significantly smaller than one, it can result in false positives.
        `atol` should be carefully selected for the use case at hand. A zero value
        for `atol` will result in `False` if either `a` or `b` is zero.
    
        Examples
        --------
        >>> np.isclose([1e10,1e-7], [1.00001e10,1e-8])
        array([True, False])
        >>> np.isclose([1e10,1e-8], [1.00001e10,1e-9])
        array([True, True])
        >>> np.isclose([1e10,1e-8], [1.0001e10,1e-9])
        array([False, True])
        >>> np.isclose([1.0, np.nan], [1.0, np.nan])
        array([True, False])
        >>> np.isclose([1.0, np.nan], [1.0, np.nan], equal_nan=True)
        array([True, True])
        >>> np.isclose([1e-8, 1e-7], [0.0, 0.0])
        array([ True, False], dtype=bool)
        >>> np.isclose([1e-100, 1e-7], [0.0, 0.0], atol=0.0)
        array([False, False], dtype=bool)
        >>> np.isclose([1e-10, 1e-10], [1e-20, 0.0])
        array([ True,  True], dtype=bool)
        >>> np.isclose([1e-10, 1e-10], [1e-20, 0.999999e-10], atol=0.0)
        array([False,  True], dtype=bool)
        """
        def within_tol(x, y, atol, rtol):
            with errstate(invalid='ignore'):
                return less_equal(abs(x-y), atol + rtol * abs(y))
    
        x = asanyarray(a)
        y = asanyarray(b)
    
        # Make sure y is an inexact type to avoid bad behavior on abs(MIN_INT).
        # This will cause casting of x later. Also, make sure to allow subclasses
        # (e.g., for numpy.ma).
>       dt = multiarray.result_type(y, 1.)
E       TypeError: invalid type promotion

../../.virtualenvs/pygbm/lib/python3.7/site-packages/numpy/core/numeric.py:2360: TypeError
```
</details>

I didn't really looked into it TBH...

Reply:
Alright then.